common/ledger/blockledger/fileledger/factory.go

```diff
@@ -26,39 +28,54 @@ type fileLedgerFactory struct {
 	mutex              sync.Mutex
 }
 
-// GetOrCreate gets an existing ledger (if it exists) or creates it if it does not
-func (flf *fileLedgerFactory) GetOrCreate(chainID string) (blockledger.ReadWriter, error) {
-	flf.mutex.Lock()
-	defer flf.mutex.Unlock()
+// GetOrCreate gets an existing ledger (if it exists) or creates it
+// if it does not.
+func (f *fileLedgerFactory) GetOrCreate(channelID string) (blockledger.ReadWriter, error) {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
 
-	key := chainID
 	// check cache
-	ledger, ok := flf.ledgers[key]
+	ledger, ok := f.ledgers[channelID]
 	if ok {
 		return ledger, nil
 	}
 	// open fresh
-	blockStore, err := flf.blkstorageProvider.Open(key)
+	blockStore, err := f.blkstorageProvider.Open(channelID)
 	if err != nil {
 		return nil, err
 	}
 	ledger = NewFileLedger(blockStore)
-	flf.ledgers[key] = ledger
+	f.ledgers[channelID] = ledger
 	return ledger, nil
 }
 
-// ChannelIDs returns the channel IDs the factory is aware of
-func (flf *fileLedgerFactory) ChannelIDs() []string {
-	channelIDs, err := flf.blkstorageProvider.List()
+// Remove removes an existing ledger.
+func (f *fileLedgerFactory) Remove(channelID string) error {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	err := f.blkstorageProvider.Drop(channelID)
+	if err != nil {
+		return err
+	}
+
+	delete(f.ledgers, channelID)
+
+	return nil
+}
```

Comment:
We need to make the `Remove` operation crash fault tolerant. The `blkstorageProvider.Drop(channelID)` may take a while, removing the index and then the folder, and if in that time there is a crash there is no recovery from `blkstorageProvider`. We need to mark the channel for removal with the file-repo at the start of this method, and unmark at the end. If we restart and find a channel marked for removal, we deed to retry the `Drop()`. 
This can be done in this PR or under a separate task, as you prefer, I think.

Reply:
Opened [FAB-18082](https://jira.hyperledger.org/browse/FAB-18082)